thinc/shims/pytorch.py

```diff
@@ -53,39 +54,13 @@ def backprop(grads):
 
         return output, backprop
 
-    def finish_update(self, optimizer):
-        pytorch_opt = self._get_pytorch_optimizer(optimizer)
-        if getattr(optimizer, "grad_clip", None):
-            torch.nn.utils.clip_grad_norm_(
-                self._model.parameters(), optimizer.grad_clip
-            )
-        pytorch_opt.step()
-        pytorch_opt.zero_grad()
-        self._update_pytorch_averages(optimizer)
-
-    def _get_pytorch_optimizer(self, sgd):
-        args = {
-            "lr": sgd.learn_rate,
-            "weight_decay": sgd.L2
-        }
-
-        if sgd.b1 != 0 and sgd.b2 != 0:
-            args["betas"] = (sgd.b1, sgd.b2)
-            args["eps"] = sgd.eps
-            if sgd.L2_is_weight_decay:
-                cls = torch.optim.AdamW
-            else:
-                cls = torch.optim.Adam
-        else:
-            cls = torch.optim.SGD
-            if sgd.b2 == 0:
-                args["momentum"] = sgd.b1
-        if self._optimizer is None:
-            self._optimizer = cls(self._model.parameters(), **args)
-        else:
-            for param_group in self._optimizer.param_groups:
-                param_group.update(args)
-        return self._optimizer
+    def finish_update(self, optimizer: Optimizer):
+        for name, value in self._model.named_parameters():
+            cpu_data = cast(FloatsXd, torch2xp(value.data))
```

Comment:
Hm, `cpu_` seems like a misnomer, right? We're not copying between devices, we're just changing the container. We can call it `xp_data`, and maybe make the loop variable `torch_data`? Alternatively I've been using `array` for the numpy/cupy typed data and `tensor` for the pytorch-typed data, but that's maybe confusing.

Reply:
Nice catch, you're right! I originally had this as `value.data.cpu().numpy()` before switching to torch2xp. xp_data sounds more appropriate now